src/Microsoft.OData.Edm/Validation/ExpressionTypeChecker.cs

```diff
@@ -486,6 +488,31 @@ private static bool TryCastBinaryConstantAsType(IEdmBinaryConstantExpression exp
             return true;
         }
 
+        private static bool TryCastEnumConstantAsType(IEdmEnumMemberExpression expression, IEdmTypeReference type, bool matchExactly, out IEnumerable<EdmError> discoveredErrors)
+        {
+            if (!type.IsEnum())
+            {
+                discoveredErrors = new EdmError[]
+                {
+                    new EdmError(expression.Location(),
+                    EdmErrorCode.ExpressionEnumKindNotValidForAssertedType,
+                    Edm.Strings.EdmModel_Validator_Semantic_ExpressionEnumKindNotValidForAssertedType)
+                };
+                return false;
+            }
+
+            foreach (var member in expression.EnumMembers)

```

Comment:
>expression [](start = 35, length = 10)

Is there any chance that expression may be null? Should we at least assert, if not check for null?